Qq/Commands.lean

```diff
@@ -0,0 +1,97 @@
+import Qq.Macro
+import Lean
+
+namespace Qq
+open Lean Meta Elab Tactic
+
+private
+def mkHaveFVars (assignments : Array Expr) (body : Expr) : MetaM Expr := do
+  let ctx ← getLCtx
+  let (_, res) ← ctx.foldrM (init := (assignments, body)) fun decl (assignments, body) => do
+    let value := assignments.back!
+    let result := .letE decl.userName decl.type value (body.abstract #[.fvar decl.fvarId]) false
+    return (assignments.pop, result)
+  return res
+
+/--
+`by_elabq` is the Qq analogue to `by_elab`
+Executes arbitrary `TermElabM` code in place of a term.
+Moreover, all variables available in the local context are accessible as
+quoted expressions, and the return value is Q-annotated as well.
+Example:
+```
+def f (x : Prop) [Decidable x] : Int :=
+  by_elabq
+    Lean.logInfo x
+    Lean.logInfo x.ty
+    return q(if $x then 2 else 3)
+```
+See also: `run_tacq`.
+-/
+scoped
+elab "by_elabq" e:doSeq : term <= expectedType => do
+  let lctx ← getLCtx
+  let levelNames ← Term.getLevelNames
+  let (quotedCtx, assignments, quotedGoal) ← liftMetaM <| StateT.run'
+        (s := {mayPostpone := False}) do
+    let (quotedCtx, assignments) ← Impl.quoteLCtx lctx levelNames
+    let quotedGoal : Expr ←
+      if expectedType.isMVar then pure q(TermElabM Expr)
+      else pure <| mkApp (mkConst ``TermElabM) (mkApp (mkConst ``Quoted)
+        (← Qq.Impl.quoteExpr expectedType))
+    return (quotedCtx, assignments, quotedGoal)
+  let codeExpr : Expr ← withLCtx quotedCtx #[] do
+    let body ← Term.elabTermAndSynthesize (← `(do $e)) quotedGoal
+    mkHaveFVars assignments body
+  let typeExpr := q(TermElabM Expr)
+  let code ← unsafe evalExpr (TermElabM Expr) typeExpr codeExpr
+  code
+
+/--
+`run_tacq` is a Qq-analogue to `run_tac`
+Executes an arbitrary `TacticM` code. Moreover, the local
+context can be directly accessed as quoted expressions,
+and the Q-annotated goal is available as `goalq`.
+Example:
+```
+example (a b : Nat) (h : a = b) : True := by
+  run_tacq
+    let p : Q(Prop) := q($a = $b)
+    let t ← Lean.Meta.inferType h
+    Lean.logInfo p
+    Lean.logInfo <| toString (← Lean.Meta.isDefEq t p)
+    Lean.logInfo <| toString (← Lean.Meta.isDefEq h.ty p)
+    Lean.logInfo goalq
+    Lean.logInfo goalq.ty
+  trivial
+```
+See also: `by_elabq`.
+-/
+scoped
+elab "run_tacq" e:doSeq : tactic => do
+  let goal ← try
+    getMainGoal
+  catch _ =>
+    throwError "no open goal, run_tacq requires main goal"
+  withMainContext do
+    let goalName := `goalq
```

Comment:
I don't like the hygiene here; I think the syntax should be `"run_tacq" goalId:ident "=>" e:doSeq`, where `goalId` is used in place of the constant name `goalq`.

You could also support a version without the `=>` that doesn't do any goal processing.